src/librustc_codegen_ssa/mir/place.rs

```diff
@@ -396,126 +396,129 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
-                        );
-                        let index = self.codegen_operand(bx, index);
-                        let llindex = index.immediate();
-                        cg_base.project_index(bx, llindex)
-                    }
-                    mir::ProjectionElem::ConstantIndex { offset,
-                                                         from_end: false,
-                                                         min_length: _ } => {
-                        let lloffset = bx.cx().const_usize(offset as u64);
-                        cg_base.project_index(bx, lloffset)
-                    }
-                    mir::ProjectionElem::ConstantIndex { offset,
-                                                         from_end: true,
-                                                         min_length: _ } => {
-                        let lloffset = bx.cx().const_usize(offset as u64);
-                        let lllen = cg_base.len(bx.cx());
-                        let llindex = bx.sub(lllen, lloffset);
-                        cg_base.project_index(bx, llindex)
-                    }
-                    mir::ProjectionElem::Subslice { from, to } => {
-                        let mut subslice = cg_base.project_index(bx,
-                            bx.cx().const_usize(from as u64));
-                        let projected_ty = PlaceTy::from_ty(cg_base.layout.ty)
-                            .projection_ty(tcx, &projection.elem).ty;
-                        subslice.layout = bx.cx().layout_of(self.monomorphize(&projected_ty));
-
-                        if subslice.layout.is_unsized() {
-                            subslice.llextra = Some(bx.sub(cg_base.llextra.unwrap(),
-                                bx.cx().const_usize((from as u64) + (to as u64))));
-                        }
-
-                        // Cast the place pointer type to the new
-                        // array or slice type (*[%_; new_len]).
-                        subslice.llval = bx.pointercast(subslice.llval,
-                            bx.cx().type_ptr_to(bx.cx().backend_type(subslice.layout)));
+                ) => {
+                    // NB: The layout of a static may be unsized as is the case when working
+                    // with a static that is an extern_type.
+                    let layout = cx.layout_of(self.monomorphize(&ty));
+                    let static_ = bx.get_static(*def_id);
+                    Some(PlaceRef::new_thin_place(bx, static_, layout, layout.align.abi))
+                }
+            };
 
-                        subslice
+            for proj in place_projection {
+                match proj {
+                    mir::Projection {
+                        ref base,
+                        elem: mir::ProjectionElem::Deref
+                    } => {
+                        // Load the pointer from its location.
+                        cg_base = Some(self.codegen_consume(bx, base).deref(bx.cx()));
```

Comment:
I think this is part of the problem (in addition to the `break` below).

So... let's assume we have `*(*foo).bar`, so our projections are `[deref, field, deref]`. What happens right now is that we evaluate `foo` (above, in the `PlaceBase` code). Then you get to `deref` (here) and call `codegen_consume` on `foo` (which is in the `base` field). Then you break, and skip the `bar` field projection and the second deref.

it will probably just work out by removing the break, but we'd still  be computing the base twice. I think we can do better, but may have to wait until we move to slices for projections

Reply:
Done and added a FIXME also

Reply:
I was just thinking, if we have `[field, *, field, *, field, *, field, *]` as projections, then we throw away 4 different base computations. I think we should resolve this now, and we can do that by iterating backwards until the first `Deref` (if any) and call `codegen_base` on it (or just compute the base if we didn't encounter one) and then iterate forward again from the found deref (skipping it). The forward iteration can then `bug!` in case it sees a `Deref` again